### ISSUE TITLE
Deleting schemas only results in soft deletion of schemas

### ISSUE THREAD

**Issue by bluedog13** [confluentinc/terraform-provider-confluent#179]
When we create and delete a schema using Terraform, it results in the "soft deletion" of the schema. 

**This is problematic for us, since the "soft deleted" schemas also count towards the "Free Schemas Limit" quota for the "Stream Governance Package" that we have.** 

Will it be possible to do a "permanent delete" on the schema,  something that is currently supported in the CLI with the  "--permanent" flag? Details can be found [here](https://docs.confluent.io/confluent-cli/current/command-reference/schema-registry/schema/confluent_schema-registry_schema_delete.html#confluent-schema-registry-schema-delete).

Thanks in advance.

**Comment by S1M0NM** [comment]
@linouk23 I'll answer your question from #124 here

I took a look at the Confluent Cloud UI and it is currently not possible to perform a hard delete for schemas there. Only a soft delete is possible.

Especially since an "immediate hard-delete" can have serious consequences if it is carried out incorrectly, because the clients use the schema ID to serialize the data. After a soft delete, the data serialized with it can still be deserialized.

The [Guidelines] (https://docs.confluent.io/platform/current/schema-registry/schema-deletion-guidelines.html#guidelines-for-using-these-apis) give a few reasons when it might be useful to use the API.

I definitely agree that there needs to be a way to also delete soft-deleted schemas, but perhaps you should do this carefully via the CLI or with curl, since you definitely have no way of then deserializing the data again if you hard deleted it.

Another idea, which would be a feature request for the schema registry, might be to introduce a config that allows defining a point in time after which soft-deleted schemas are permanently deleted.

**Comment by linouk23** [comment]
Thank you @S1M0NM! Given the feedback from https://github.com/confluentinc/terraform-provider-confluent/issues/124#issuecomment-1403584887 and https://github.com/confluentinc/terraform-provider-confluent/issues/179#issue-1556865092 it looks like it may be a good idea to perform a soft deletion by default but have an option to override if necessary. In other words, we're looking at a boolean flag `enable_hard_deletion` or something that defaults to `true`. How does it sound to you @S1M0NM @bluedog13?

**Comment by bluedog13** [comment]
Thanks. 

As long as we have the ability to perform a hard delete (since it's supported in the CLI and we use it to clean up), it works.  Having a flag would also mimic the CLI behavior.

**Comment by S1M0NM** [comment]
@linouk23 The flag is a good idea.

(sorry forgot that point in my comment)

**Comment by S1M0NM** [comment]
I was just wondering whether a `soft_delete` flag would be possible, which is `false` by default.
Setting `soft_delete` to `true` will update the `confluent_schema` resource, which will do a soft delete on the schema, and removing the resource entirely will do a hard delete.

---

Another point, which is still a question for me, is how enable_hard_deletion behaves in connection with a schema update. Does this mean that if I update the schema attribute of the confluent_schema resource, the current one will be evolved and the previous version will be hard deleted?

cc @linouk23 @bluedog13

**Comment by linouk23** [comment]
> I was just wondering whether a soft_delete flag would be possible, which is false by default.

Sure. I'll take a look at naming conventions for naming boolean attributes so we could slightly rename it but yeah overall it sounds good to me.

> Another point, which is still a question for me, is how enable_hard_deletion behaves in connection with a schema update. Does this mean that if I update the schema attribute of the confluent_schema resource, the current one will be evolved and the previous version will be hard deleted?

So there're 2 options:

* if `enable_updates` is set to `true` then all previous version will still exist and there's no deletions at all.
* if `enable_updates` is set to `false` then schemas will be hard / soft deleted (depends on the value of `soft_delete` flag).

How does it sound to you? cc @S1M0NM

**Comment by linouk23** [comment]
@S1M0NM 

> I was just wondering whether a soft_delete flag would be possible, which is false by default.
Setting soft_delete to true will update the confluent_schema resource, which will do a soft delete on the schema, and removing the resource entirely will do a hard delete.

It looks like

> I took a look at the Confluent Cloud UI and it is currently not possible to perform a hard delete for schemas there. Only a soft delete is possible.
> Especially since an "immediate hard-delete" can have serious consequences if it is carried out incorrectly, because the clients use the schema ID to serialize the data. After a soft delete, the data serialized with it can still be deserialized.
> The [Guidelines](https://docs.confluent.io/platform/current/schema-registry/schema-deletion-guidelines.html#guidelines-for-using-these-apis) give a few reasons when it might be useful to use the API.
> I definitely agree that there needs to be a way to also delete soft-deleted schemas, but perhaps you should do this carefully via the CLI or with curl, since you definitely have no way of then deserializing the data again if you hard deleted it.
> Another idea, which would be a feature request for the schema registry, might be to introduce a config that allows defining a point in time after which soft-deleted schemas are permanently deleted.

implies that soft deletion by default is expected here so I'm thinking about `enable_hard_deletion = false` by default.

> Setting soft_delete to true will update the confluent_schema resource, which will do a soft delete on the schema, and removing the resource entirely will do a hard delete.

Regarding this -- I think it's going to be a 2 step process: first just `enable_hard_deletion` is updated to `true` and then the next step is to remove a resource.

**Comment by bluedog13** [comment]
I prefer having "soft delete" as default and a flag for hard deletion (false as default like you mention above). 

This keeps it inline with how we use CLI and will not causing confusion between folks who use both CLI and Terraform for automation.

**Comment by S1M0NM** [comment]
> Regarding this -- I think it's going to be a 2 step process: first just `enable_hard_deletion` is updated to `true` and then the next step is to remove a resource.

What would the workflow look like to first perform a soft delete on a schema to restrict its use for serializing new data and then completely (hard) delete it later after x days because there is no longer any data serialized with it?

> What about hard_deletion which is false by default?

sounds good to me

**Comment by dylanmei** [comment]
It should not need to be 2 step process. This is reasonable, and makes sense in CI/CD environments:

```
hard_delete = var.enviroment == "dev" ? true : false
```

**Comment by linouk23** [comment]
@dylanmei hmm I'll run an experiment but I'm pretty sure it'd have to be 2 step process -- see [this](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/bigquery_table) 
![image](https://user-images.githubusercontent.com/5459870/215183223-6f514099-1a95-4d7f-8458-0c0af09e785e.png)

More specifically, there's a scenario where a user sets
```
hard_delete = true
```
and then comments out a resource to delete instead of running a terraform destroy -- this way TF Provider won't be able to read an updated value for `hard_delete` property (since it was commented otu) so it'd be using `false` (a value before an update) when deleting a resource wdyt.

**Comment by linouk23** [comment]
> What would the workflow look like to first perform a soft delete on a schema to restrict its use for serializing new data and then completely (hard) delete it later after x days because there is no longer any data serialized with it?

Hmm I'm not super sure so I might need your help here but it looks like first they'd delete it from TF config (by using a default value of `hard_delete`) and then they could go through [our docs](https://docs.confluent.io/platform/current/schema-registry/schema-deletion-guidelines.html#hard-delete-a-schema) that describe how to run a hard delete via API (and CLI?)? Does it make sense to you @S1M0NM?

**Comment by S1M0NM** [comment]
> Hmm I'm not super sure so I might need your help here but it looks like first they'd delete it from TF config (by using a default value of `hard_delete`) and then they could go through [our docs](https://docs.confluent.io/platform/current/schema-registry/schema-deletion-guidelines.html#hard-delete-a-schema) that describe how to run a hard delete via API (and CLI?)? Does it make sense to you @S1M0NM?

Hmm.. I'm not entirely sure about that. 
One of my ideas went along the lines of a `soft_delete` flag, which is `false` by default.
Deleting the `confluent_schema` resource would always lead to a hard deletion of the schema.
If I want to soft delete the schema first, I set `soft_delete` to `true` and after a certain period of time I delete the entire resource from my Terraform Config, which then performs the hard deletion.

However, this procedure could have dangerous side effects if the deletion happens accidentally and you carry out the hard deletion as a result.
The idea of cleaning up yourself via the API could then be safer

**Comment by linouk23** [comment]
I'm happy to announce that we've just released a new `1.27.0` [version](https://github.com/confluentinc/terraform-provider-confluent/blob/master/CHANGELOG.md#1270-january-30-2023) of TF Provider that supports `hard_delete` attribute (defaults to `false`). Find more details in the [docs](https://registry.terraform.io/providers/confluentinc/confluent/latest/docs/resources/confluent_schema):
![image](https://user-images.githubusercontent.com/5459870/215622786-f25448a0-5127-43f0-9fb1-085572a88bf2.png)


cc @bluedog13 @S1M0NM
